### ISSUE TITLE
mmdetection getting validation loss after each epoch of training

### ISSUE THREAD

**Issue by ekaraali** [open-mmlab/mmdetection#7479]
Hello,

We are training Faster R-CNN model for object detection. We would like to make hyperparameter tuning using ray tune. Therefore, we need validation losses after each epoch. We have used workflow = [('train', 1),('val', 1)] in default_runtime.py file, but it writes the losses into log.json file. How can we get losses before it saves them into the log.json file as variable so that we can give this variable to ray tune function.

**Comment by Czm369** [comment]
Maybe you can write a hook function based on `after_train_epoch`, which is in [hooks](https://github.com/open-mmlab/mmcv/tree/master/mmcv/runner/hooks)/hook.py

**Comment by tanlei0** [comment]
My solution code as follows, it does some extra training. 
You need modify your config file to match this code. 
The log will save at the work-dir.
May it can help u.

```
import mmcv
import torch
from mmcv.runner import HOOKS, Hook
from mmdet.core.evaluation.eval_hooks import EvalHook
from mmcv import Config
from mmdet.datasets import (build_dataloader, build_dataset)
from collections import OrderedDict
import datetime
import os.path as osp

@HOOKS.register_module()
class ValLossHook(Hook):
    def __init__(self, cfg_file):
        super(ValLossHook).__init__()   
        cfg = Config.fromfile(cfg_file)
        val_samples_per_gpu = 1
        cfg.data.val["pipeline"].insert(1, {'type': 'LoadAnnotations', 'with_bbox': True, 'with_mask': True})
        cfg.data.val["pipeline"].pop()
        cfg.data.val["pipeline"].append({'type': 'RandomFlip', 'flip_ratio': 0.5})
        cfg.data.val["pipeline"].append({'type': 'Normalize', 'mean': [123.675, 116.28, 103.53], 'std': [58.395, 57.12, 57.375], 'to_rgb': True})
        cfg.data.val["pipeline"].append({'type': 'Pad', 'size_divisor': 32})
        cfg.data.val["pipeline"].append({'type': 'DefaultFormatBundle'})
        cfg.data.val["pipeline"].append({'type': 'Collect', 'keys': ['img', 'gt_bboxes', 'gt_labels', 'gt_masks']})
        
        """
        # some test code
        cfg.data.val["pipeline"][-1]["transforms"][-1]["keys"] = ['img', 'gt_bboxes', 'gt_labels', 'gt_masks']
        cfg.data.val["pipeline"][-1]["transforms"][-2]["keys"] = ['img', 'gt_bboxes', 'gt_labels', 'gt_masks']
        
        
        {'type': 'LoadImageFromFile'}
        {'type': 'LoadAnnotations', 'with_bbox': True, 'with_mask': True}
        {'type': 'Normalize', 'mean': [123.675, 116.28, 103.53], 'std': [58.395, 57.12, 57.375], 'to_rgb': True}
        {'type': 'Pad', 'size_divisor': 32}
        {'type': 'DefaultFormatBundle'}
        {'type': 'Collect', 'keys': ['img', 'gt_bboxes', 'gt_labels', 'gt_masks']}
        """
        val_dataset = build_dataset(cfg.data.val)
        val_dataloader = build_dataloader(
            val_dataset,
            samples_per_gpu=val_samples_per_gpu,
            workers_per_gpu=cfg.data.workers_per_gpu,
            dist=False,
            shuffle=False)
        
        self.dataset = val_dataset
        self.dataloader = val_dataloader

    def after_epoch(self, runner):
        self.json_log_path = osp.join(runner.work_dir,
                                      f'{runner.timestamp}.valloss.json')
        self._do_evaluate(runner)


    def _do_evaluate(self, runner):
        """perform evaluation and save ckpt."""

        model = runner.model
        model.eval()
        results = []

        """
        # find eval hook in runner
        for hook in runner.hooks:
            if isinstance(hook, EvalHook):
                ehook = hook
                break
        else:
            raise("No Eval Hook")
        dataloader = ehook.dataloader
        dataset = dataloader.dataset
        """

        prog_bar = mmcv.ProgressBar(len(self.dataset))
        for i, data in enumerate(self.dataloader):
            with torch.no_grad():
                outputs = runner.model.train_step(data, runner.optimizer)

        
            results.extend(outputs)
            prog_bar.update()

            if 'log_vars' in outputs:
                runner.log_buffer.update(outputs['log_vars'], outputs['num_samples'])

        runner.log_buffer.average(outputs['num_samples'])
        if runner.log_buffer.ready:
            self.log(runner)
            runner.log_buffer.clear_output()

            

    def log(self, runner):
        if 'eval_iter_num' in runner.log_buffer.output:
            # this doesn't modify runner.iter and is regardless of by_epoch
            cur_iter = runner.log_buffer.output.pop('eval_iter_num')
        else:
            cur_iter = runner.iter

        log_dict = OrderedDict(
            mode="EVAL",
            epoch=runner.epoch + 1,
            iter=cur_iter)

        # only record lr of the first param group
        cur_lr = runner.current_lr()
        if isinstance(cur_lr, list):
            log_dict['lr'] = cur_lr[0]
        else:
            assert isinstance(cur_lr, dict)
            log_dict['lr'] = {}
            for k, lr_ in cur_lr.items():
                assert isinstance(lr_, list)
                log_dict['lr'].update({k: lr_[0]})

        if 'time' in runner.log_buffer.output:
            # statistic memory
            if torch.cuda.is_available():
                log_dict['memory'] = None

        log_dict = dict(log_dict, **runner.log_buffer.output)

        self._log_info(log_dict, runner)
        self._dump_log(log_dict, runner)
        return log_dict

    def _log_info(self, log_dict, runner):
        # print exp name for users to distinguish experiments
        # at every ``interval_exp_name`` iterations and the end of each epoch

        # val/test time
        # here 1000 is the length of the val dataloader
        # by epoch: Epoch[val] [4][1000]
        # by iter: Iter[val] [1000]
        if True:
            log_str = f'Epoch({log_dict["mode"]}) ' \
                f'[{log_dict["epoch"]}][{log_dict["iter"]}]\t'


        log_items = []
        for name, val in log_dict.items():
            # TODO: resolve this hack
            # these items have been in log_str
            if name in [
                    'mode', 'Epoch', 'iter', 'lr', 'time', 'data_time',
                    'memory', 'epoch'
            ]:
                continue
            if isinstance(val, float):
                val = f'{val:.4f}'
            log_items.append(f'{name}: {val}')
        log_str += ', '.join(log_items)

        runner.logger.info(log_str)

    def _round_float(self, items):
        if isinstance(items, list):
            return [self._round_float(item) for item in items]
        elif isinstance(items, float):
            return round(items, 5)
        else:
            return items

    def _dump_log(self, log_dict, runner):
        # dump log in json format
        json_log = OrderedDict()
        for k, v in log_dict.items():
            json_log[k] = self._round_float(v)
        # only append log at last line
        if runner.rank == 0:
            with open(self.json_log_path, 'a+') as f:
                mmcv.dump(json_log, f, file_format='json')
                f.write('\n')
```

**Comment by hoya-cho** [comment]
> My solution code as follows, it does some extra training. You need modify your config file to match this code. The log will save at the work-dir. May it can help u.
> 
> ```
> import mmcv
> import torch
> from mmcv.runner import HOOKS, Hook
> from mmdet.core.evaluation.eval_hooks import EvalHook
> from mmcv import Config
> from mmdet.datasets import (build_dataloader, build_dataset)
> from collections import OrderedDict
> import datetime
> import os.path as osp
> 
> @HOOKS.register_module()
> class ValLossHook(Hook):
>     def __init__(self, cfg_file):
>         super(ValLossHook).__init__()   
>         cfg = Config.fromfile(cfg_file)
>         val_samples_per_gpu = 1
>         cfg.data.val["pipeline"].insert(1, {'type': 'LoadAnnotations', 'with_bbox': True, 'with_mask': True})
>         cfg.data.val["pipeline"].pop()
>         cfg.data.val["pipeline"].append({'type': 'RandomFlip', 'flip_ratio': 0.5})
>         cfg.data.val["pipeline"].append({'type': 'Normalize', 'mean': [123.675, 116.28, 103.53], 'std': [58.395, 57.12, 57.375], 'to_rgb': True})
>         cfg.data.val["pipeline"].append({'type': 'Pad', 'size_divisor': 32})
>         cfg.data.val["pipeline"].append({'type': 'DefaultFormatBundle'})
>         cfg.data.val["pipeline"].append({'type': 'Collect', 'keys': ['img', 'gt_bboxes', 'gt_labels', 'gt_masks']})
>         
>         """
>         # some test code
>         cfg.data.val["pipeline"][-1]["transforms"][-1]["keys"] = ['img', 'gt_bboxes', 'gt_labels', 'gt_masks']
>         cfg.data.val["pipeline"][-1]["transforms"][-2]["keys"] = ['img', 'gt_bboxes', 'gt_labels', 'gt_masks']
>         
>         
>         {'type': 'LoadImageFromFile'}
>         {'type': 'LoadAnnotations', 'with_bbox': True, 'with_mask': True}
>         {'type': 'Normalize', 'mean': [123.675, 116.28, 103.53], 'std': [58.395, 57.12, 57.375], 'to_rgb': True}
>         {'type': 'Pad', 'size_divisor': 32}
>         {'type': 'DefaultFormatBundle'}
>         {'type': 'Collect', 'keys': ['img', 'gt_bboxes', 'gt_labels', 'gt_masks']}
>         """
>         val_dataset = build_dataset(cfg.data.val)
>         val_dataloader = build_dataloader(
>             val_dataset,
>             samples_per_gpu=val_samples_per_gpu,
>             workers_per_gpu=cfg.data.workers_per_gpu,
>             dist=False,
>             shuffle=False)
>         
>         self.dataset = val_dataset
>         self.dataloader = val_dataloader
> 
>     def after_epoch(self, runner):
>         self.json_log_path = osp.join(runner.work_dir,
>                                       f'{runner.timestamp}.valloss.json')
>         self._do_evaluate(runner)
> 
> 
>     def _do_evaluate(self, runner):
>         """perform evaluation and save ckpt."""
> 
>         model = runner.model
>         model.eval()
>         results = []
> 
>         """
>         # find eval hook in runner
>         for hook in runner.hooks:
>             if isinstance(hook, EvalHook):
>                 ehook = hook
>                 break
>         else:
>             raise("No Eval Hook")
>         dataloader = ehook.dataloader
>         dataset = dataloader.dataset
>         """
> 
>         prog_bar = mmcv.ProgressBar(len(self.dataset))
>         for i, data in enumerate(self.dataloader):
>             with torch.no_grad():
>                 outputs = runner.model.train_step(data, runner.optimizer)
> 
>         
>             results.extend(outputs)
>             prog_bar.update()
> 
>             if 'log_vars' in outputs:
>                 runner.log_buffer.update(outputs['log_vars'], outputs['num_samples'])
> 
>         runner.log_buffer.average(outputs['num_samples'])
>         if runner.log_buffer.ready:
>             self.log(runner)
>             runner.log_buffer.clear_output()
> 
>             
> 
>     def log(self, runner):
>         if 'eval_iter_num' in runner.log_buffer.output:
>             # this doesn't modify runner.iter and is regardless of by_epoch
>             cur_iter = runner.log_buffer.output.pop('eval_iter_num')
>         else:
>             cur_iter = runner.iter
> 
>         log_dict = OrderedDict(
>             mode="EVAL",
>             epoch=runner.epoch + 1,
>             iter=cur_iter)
> 
>         # only record lr of the first param group
>         cur_lr = runner.current_lr()
>         if isinstance(cur_lr, list):
>             log_dict['lr'] = cur_lr[0]
>         else:
>             assert isinstance(cur_lr, dict)
>             log_dict['lr'] = {}
>             for k, lr_ in cur_lr.items():
>                 assert isinstance(lr_, list)
>                 log_dict['lr'].update({k: lr_[0]})
> 
>         if 'time' in runner.log_buffer.output:
>             # statistic memory
>             if torch.cuda.is_available():
>                 log_dict['memory'] = None
> 
>         log_dict = dict(log_dict, **runner.log_buffer.output)
> 
>         self._log_info(log_dict, runner)
>         self._dump_log(log_dict, runner)
>         return log_dict
> 
>     def _log_info(self, log_dict, runner):
>         # print exp name for users to distinguish experiments
>         # at every ``interval_exp_name`` iterations and the end of each epoch
> 
>         # val/test time
>         # here 1000 is the length of the val dataloader
>         # by epoch: Epoch[val] [4][1000]
>         # by iter: Iter[val] [1000]
>         if True:
>             log_str = f'Epoch({log_dict["mode"]}) ' \
>                 f'[{log_dict["epoch"]}][{log_dict["iter"]}]\t'
> 
> 
>         log_items = []
>         for name, val in log_dict.items():
>             # TODO: resolve this hack
>             # these items have been in log_str
>             if name in [
>                     'mode', 'Epoch', 'iter', 'lr', 'time', 'data_time',
>                     'memory', 'epoch'
>             ]:
>                 continue
>             if isinstance(val, float):
>                 val = f'{val:.4f}'
>             log_items.append(f'{name}: {val}')
>         log_str += ', '.join(log_items)
> 
>         runner.logger.info(log_str)
> 
>     def _round_float(self, items):
>         if isinstance(items, list):
>             return [self._round_float(item) for item in items]
>         elif isinstance(items, float):
>             return round(items, 5)
>         else:
>             return items
> 
>     def _dump_log(self, log_dict, runner):
>         # dump log in json format
>         json_log = OrderedDict()
>         for k, v in log_dict.items():
>             json_log[k] = self._round_float(v)
>         # only append log at last line
>         if runner.rank == 0:
>             with open(self.json_log_path, 'a+') as f:
>                 mmcv.dump(json_log, f, file_format='json')
>                 f.write('\n')
> ```

Where do I use the ValLossHook Class I implemented?
For example, is it used in eval_hooks.py or single_gpu_test , multi_gpu_test methods in test.py ?

May I know the example you used?

**Comment by BIGWangYuDong** [comment]
You just need to set workflow = [('train', 1),('val', 1)] in the configuration
